README.org

```diff
@@ -240,12 +240,20 @@ The following snippet will refresh caches when one of the bib files change:
 #+BEGIN_SRC emacs-lisp
 (bibtex-actions-with-filenotify-global #'bibtex-actions-refresh)
 
-(add-hook 'LaTeX-mode-hook (lambda () (bibtex-actions-with-filenotify-local #'bibtex-actions-refresh)))
-(add-hook 'org-mode-hook (lambda () (bibtex-actions-with-filenotify-local #'bibtex-actions-refresh)))
+(add-hook 'LaTeX-mode-hook
+          (lambda () (bibtex-actions-with-filenotify-local #'bibtex-actions-refresh)))
+(add-hook 'org-mode-hook
+          (lambda () (bibtex-actions-with-filenotify-local #'bibtex-actions-refresh)))
+
+(defun refresh-bib-file-watches ()
+  (interactive) (bibtex-actions-with-filenotify-refresh #'bibtex-actions-refresh) (bibtex-actions-refresh))
```

Comment:
```suggestion
(defun refresh-bib-file-watches () (interactive)
       (bibtex-actions-with-filenotify-refresh
        (lambda ()
          (setq bibtex-actions--local-candidates-cache 'uninitialized)))
       (setq bibtex-actions--local-candidates-cache 'uninitialized))
```

Formatting again.

I guess my question is why wouldn't we just add this function to the code?

Again: I'm just asking the question.

Reply:
I guess the question is how much flexibility we want. In my mind people might want to use these functions with different callbacks. Similarly the extra files can be added in the optional arguments. As it is we provide a building block for the refresh function but the refresh function itself depends on what the user wants.

Though another option to provide it as a default or else we can make the callback a variable with default value being  `#'bibtex-actions-refresh`.

Reply:
So how do I add a path for notes?

Reply:
> Formatting again.

I didn't notice it earlier but the two examples have been mixed up here.

Reply:
> So how do I add a path for notes?

Assuming the path is global it will be 
```lisp
(bibtex-actions-with-filenotify-global #'bibtex-actions-refresh '(PATH))

(defun refresh-bib-file-watches () (interactive)
       (bibtex-actions-with-filenotify-refresh
        #'bibtex-actions-refresh
        '(PATH))
       (bibtex-actions-refresh))
```
But it needs to be tested. Since I don't have this setup.

Reply:
I'll see if I can test it tomorrow.

Reply:
This is what I'm wanting to do, and include in the README as an example:

```elisp
(bibtex-actions-with-filenotify-global #'bibtex-actions-refresh bibtex-completion-notes-path)

(defun refresh-bib-file-watches () (interactive)
       (bibtex-actions-with-filenotify-refresh
        #'bibtex-actions-refresh
        bibtex-completion-notes-path
       (bibtex-actions-refresh))
```

But I get a type error.

Reply:
> This is what I'm wanting to do, and include in the README as an example:
> 
> ```emacs-lisp
> (bibtex-actions-with-filenotify-global #'bibtex-actions-refresh bibtex-completion-notes-path)
> 
> (defun refresh-bib-file-watches () (interactive)
>        (bibtex-actions-with-filenotify-refresh
>         #'bibtex-actions-refresh
>         bibtex-completion-notes-path
>        (bibtex-actions-refresh))
> ```
> 
> But I get a type error.

The arguments that they expect are a list of paths so try this:

```lisp
(bibtex-actions-with-filenotify-global #'bibtex-actions-refresh '(bibtex-completion-notes-path))

(defun refresh-bib-file-watches () (interactive)
       (bibtex-actions-with-filenotify-refresh
        #'bibtex-actions-refresh
        '(bibtex-completion-notes-path)
       (bibtex-actions-refresh))
```

Reply:
I should have mentioned that I did try that. Still get a "126" type error.

Reply:
I am at a loss, I added this to my config,

```lisp
(bibtex-actions-with-filenotify-global #'bibtex-actions-refresh '("~/Downloads"))

(defun refresh-bib-files () "Rescan the bib file and add updated watches" (interactive)
       (bibtex-actions-with-filenotify-refresh #'bibtex-actions-refresh '("~/Downloads")) (bibtex-actions-refresh))
``` 
And I see a watch on my downloads directory in `file-notify-descriptors` as expected. What is the value of `bibtex-completion-notes-path` ?

Reply:
It's a string directory path, as with you. 🤷‍♂️ 

I'll try again.

Reply:
Here's the stack trace; not sure what to make of it:

```elisp
Debugger entered--Lisp error: (wrong-type-argument 126)
  file-notify-add-watch(126 (change) (closure ((bibfile . 126) (extra-files bibtex-completion-notes-path) (func . bibtex-actions-refresh) embark-meta-map embark-general-map embark-target-finders embark-keymap-alist t) (x) (if (eq 'stopped (car (cdr x))) nil (funcall func))))
  (closure ((extra-files bibtex-completion-notes-path) (func . bibtex-actions-refresh) embark-meta-map embark-general-map embark-target-finders embark-keymap-alist t) (bibfile) (file-notify-add-watch bibfile '(change) #'(lambda (x) (if (eq 'stopped (car ...)) nil (funcall func)))))(126)
  mapcar((closure ((extra-files bibtex-completion-notes-path) (func . bibtex-actions-refresh) embark-meta-map embark-general-map embark-target-finders embark-keymap-alist t) (bibfile) (file-notify-add-watch bibfile '(change) #'(lambda (x) (if (eq 'stopped (car ...)) nil (funcall func))))) (126 47 111 114 103 47 98 105 98 47 110 101 119 101 114 46 98 105 98 bibtex-completion-notes-path))
  #f(compiled-function #'sequence #<bytecode 0x1856c9cd53e62674>)((closure ((extra-files bibtex-completion-notes-path) (func . bibtex-actions-refresh) embark-meta-map embark-general-map embark-target-finders embark-keymap-alist t) (bibfile) (file-notify-add-watch bibfile '(change) #'(lambda (x) (if (eq 'stopped (car ...)) nil (funcall func))))) (126 47 111 114 103 47 98 105 98 47 110 101 119 101 114 46 98 105 98 bibtex-completion-notes-path))
  apply(#f(compiled-function #'sequence #<bytecode 0x1856c9cd53e62674>) (closure ((extra-files bibtex-completion-notes-path) (func . bibtex-actions-refresh) embark-meta-map embark-general-map embark-target-finders embark-keymap-alist t) (bibfile) (file-notify-add-watch bibfile '(change) #'(lambda (x) (if (eq 'stopped (car ...)) nil (funcall func))))) (126 47 111 114 103 47 98 105 98 47 110 101 119 101 114 46 98 105 98 bibtex-completion-notes-path) nil)
  seq-map((closure ((extra-files bibtex-completion-notes-path) (func . bibtex-actions-refresh) embark-meta-map embark-general-map embark-target-finders embark-keymap-alist t) (bibfile) (file-notify-add-watch bibfile '(change) #'(lambda (x) (if (eq 'stopped (car ...)) nil (funcall func))))) (126 47 111 114 103 47 98 105 98 47 110 101 119 101 114 46 98 105 98 bibtex-completion-notes-path))
  (setq bibtex-actions--global-watches (seq-map #'(lambda (bibfile) (file-notify-add-watch bibfile '(change) #'(lambda (x) (if ... nil ...)))) (seq-concatenate 'list bibtex-completion-bibliography extra-files)))
  bibtex-actions-with-filenotify-global(bibtex-actions-refresh (bibtex-completion-notes-path))
  eval-region(2090 2391)  ; Reading at buffer position 2090
  funcall-interactively(eval-region 2090 2391)
  command-execute(eval-region record)
  execute-extended-command(nil "eval-region" nil)
  funcall-interactively(execute-extended-command nil "eval-region" nil)
  command-execute(execute-extended-command)
```

Reply:
> Here's the stack trace; not sure what to make of it:
> 
> ```emacs-lisp
> Debugger entered--Lisp error: (wrong-type-argument 126)
>   file-notify-add-watch(126 (change) (closure ((bibfile . 126) (extra-files bibtex-completion-notes-path) (func . bibtex-actions-refresh) embark-meta-map embark-general-map embark-target-finders embark-keymap-alist t) (x) (if (eq 'stopped (car (cdr x))) nil (funcall func))))
>   (closure ((extra-files bibtex-completion-notes-path) (func . bibtex-actions-refresh) embark-meta-map embark-general-map embark-target-finders embark-keymap-alist t) (bibfile) (file-notify-add-watch bibfile '(change) #'(lambda (x) (if (eq 'stopped (car ...)) nil (funcall func)))))(126)
>   mapcar((closure ((extra-files bibtex-completion-notes-path) (func . bibtex-actions-refresh) embark-meta-map embark-general-map embark-target-finders embark-keymap-alist t) (bibfile) (file-notify-add-watch bibfile '(change) #'(lambda (x) (if (eq 'stopped (car ...)) nil (funcall func))))) (126 47 111 114 103 47 98 105 98 47 110 101 119 101 114 46 98 105 98 bibtex-completion-notes-path))
>   #f(compiled-function #'sequence #<bytecode 0x1856c9cd53e62674>)((closure ((extra-files bibtex-completion-notes-path) (func . bibtex-actions-refresh) embark-meta-map embark-general-map embark-target-finders embark-keymap-alist t) (bibfile) (file-notify-add-watch bibfile '(change) #'(lambda (x) (if (eq 'stopped (car ...)) nil (funcall func))))) (126 47 111 114 103 47 98 105 98 47 110 101 119 101 114 46 98 105 98 bibtex-completion-notes-path))
>   apply(#f(compiled-function #'sequence #<bytecode 0x1856c9cd53e62674>) (closure ((extra-files bibtex-completion-notes-path) (func . bibtex-actions-refresh) embark-meta-map embark-general-map embark-target-finders embark-keymap-alist t) (bibfile) (file-notify-add-watch bibfile '(change) #'(lambda (x) (if (eq 'stopped (car ...)) nil (funcall func))))) (126 47 111 114 103 47 98 105 98 47 110 101 119 101 114 46 98 105 98 bibtex-completion-notes-path) nil)
>   seq-map((closure ((extra-files bibtex-completion-notes-path) (func . bibtex-actions-refresh) embark-meta-map embark-general-map embark-target-finders embark-keymap-alist t) (bibfile) (file-notify-add-watch bibfile '(change) #'(lambda (x) (if (eq 'stopped (car ...)) nil (funcall func))))) (126 47 111 114 103 47 98 105 98 47 110 101 119 101 114 46 98 105 98 bibtex-completion-notes-path))
>   (setq bibtex-actions--global-watches (seq-map #'(lambda (bibfile) (file-notify-add-watch bibfile '(change) #'(lambda (x) (if ... nil ...)))) (seq-concatenate 'list bibtex-completion-bibliography extra-files)))
>   bibtex-actions-with-filenotify-global(bibtex-actions-refresh (bibtex-completion-notes-path))
>   eval-region(2090 2391)  ; Reading at buffer position 2090
>   funcall-interactively(eval-region 2090 2391)
>   command-execute(eval-region record)
>   execute-extended-command(nil "eval-region" nil)
>   funcall-interactively(execute-extended-command nil "eval-region" nil)
>   command-execute(execute-extended-command)
> ```

Thanks, should be fixed now. The problem with my handling of `bibtex-completions-bibliography` which I had incorrectly assumed to be list of paths. I have now fixed it and now I am passing optional arguments through `bibtex-actions--normalize-paths` too so now they work with string paths as well as list of paths.

Reply:
Yeah, it's kind of annoying that bibtex-completion allows both string and list for these.

Reply:
> Yeah, it's kind of annoying that bibtex-completion allows both string and list for these.

I have missed types during the little elisp I have written!

Does it work now? I set the `bibtex-completions-bibliography` to a string path to test and it worked for me after the changes.

Reply:
> Does it work now?

Yes!

So to go back to https://github.com/bdarcus/bibtex-actions/pull/215#discussion_r684657924, why does one need a defun to set that?

Reply:
> So to go back to [#215 (comment)](https://github.com/bdarcus/bibtex-actions/pull/215#discussion_r684657924), why does one need a defun to set that?

Mostly because of my aversion to define extra variable. An alternative way would be get the three arguments the `bibtex-actions-with-filenotify-refresh` needs from variables say `bibtex-actions-filenotify-call-back` `bibtex-actions-filenotify-extra-files` and `bibtex-actions-filenotify-extra-local-files`. Then we can move some most of the configuration to setting these values. I actually don't have any reason against this.

Reply:
To settle the final details, we should probably think about this from a user POV

In my original idea, that I'm unsure of (mostly because of diversity of pdf handling), I had thought of a defcustom where a user could could have a choice list like:

- `notes` (use `bibtex-completion-notes-path`)
- `files (PDF)`(use `bibtex-completion-library-path`)
- `custom`

Reply:
> I'm also a little unsure of the local extra file.

I don't see a use for it either, I just went with the most general and symmetric way to do it. But I don't use emacs to manage my notes or library so I don't have any idea of what someone might do with it.

Reply:
> To settle the final details, we should probably think about this from a user POV

Since I am not using these facilities myself, I actually don't know what a good design would be so I am happy to be guided.

Reply:
Am thinking ...

Reply:
> An alternative way would be get the three arguments the `bibtex-actions-with-filenotify-refresh` needs from variables say `bibtex-actions-filenotify-call-back` `bibtex-actions-filenotify-extra-files` and `bibtex-actions-filenotify-extra-local-files`.

My impulse is we should drop the extra-local stuff, or at least not expose it ATM. Thoughts?

So if our focus for this question is just on the first two:

1. the callback; obviously default is `bibtex-actions-refresh`; what others calbacks are you imagining? If you have even one good one in mind, then that would seem to justify the defcustom with default?
2. this is more tricky, but what about something like below, with an option to specify paths directly?

```elisp
(defcustom bibtex-actions-filenotify-extra nil
  "Additional files for inotify to watch."
  :group 'bibtex-actions
  :type '(repeat choice :tag "Files"
          (const notes)
          (const library)))
```

So then a user could select them in the customize UI, or do:

```elisp
(setq bibtex-actions-filenotify-extra 'notes)
```

```elisp
(setq bibtex-actions-filenotify-extra '("foo.md" "bar.md")
```

WDYT?

cc @titaniumbones

Reply:
I'm looking for a conversation with Titus on pdf issues with inotify ....

Reply:
> > An alternative way would be get the three arguments the `bibtex-actions-with-filenotify-refresh` needs from variables say `bibtex-actions-filenotify-call-back` `bibtex-actions-filenotify-extra-files` and `bibtex-actions-filenotify-extra-local-files`.
> 
> My impulse is we should drop the extra-local stuff, or at least not expose it ATM.

That seems fine to me. 

> So if our focus for this question is just on the first two:
> 
>     1. the callback; obviously default is `bibtex-actions-refresh`; what others calbacks are you imagining? If you have even one good one in mind, then that would seem to justify the defcustom with default?

Just invalidating the cache which is other thing in the documentation. I think it makes sense in the light of Titus' comments, especially for someone who might makes frequent use of notes and might not want the pauses caused by frequent cache refreshes. Another option would be refresh but on an idle timer. More generally I think it is better to not to limit other people by our imagination and if they want to do other stuff with it they should be able to. So I would say defcustom of function type makes sense.

>     2. this is more tricky, but what about something like below, with an option to specify paths directly?
> 
> 
> ```emacs-lisp
> (defcustom bibtex-actions-filenotify-extra nil
>   "Additional files for inotify to watch."
>   :group 'bibtex-actions
>   :type '(repeat choice :tag "Files"
>           (const notes)
>           (const library)))
> ```
> 
> So then a user could select them in the customize UI, or do:
> 
> ```emacs-lisp
> (setq bibtex-actions-filenotify-extra 'notes)
> ```
> 
> ```emacs-lisp
> (setq bibtex-actions-filenotify-extra '("foo.md" "bar.md"))
> ```

This looks good to me. 

> WDYT?
> 
> cc @titaniumbones

Reply:
> [oantolin/embark#176 (comment)](https://github.com/oantolin/embark/issues/176#issuecomment-817798025)

This is taken care of. The callback isn't directly passed to filenotify. We actually use the extra argument to make sure that there is no refresh when a watch is removed. Though I think it will make sense to expose this argument to the end user.

Reply:
Oh, one thing I thought of but didn't do is to make local and global caches independently refreshable. Not for the user facing code like `bibtex-actions-refresh` but for internal code like this. Since we which cache needs to be refreshed, it would be straightforward and make making them happen faster for those who use both global and local bibs. Though it would be better if such a user could comment here.

Reply:
So that would just mean a non-interactive `bibtex-actions--refresh-local` function?

Seems reasonable.

Reply:
Yes and also a `bibtex-actions--refresh-globall`. `bibtex-actions-refresh` then calls the both.

Reply:
Just occurs to.me: if we have the defcustom for the "extra" does it make sense for it to include `'bibliography` as an option, so that if chosen, it's then automatically updated?

So then the defcustom would remove "extra" from the name?

Reply:
> Just occurs to.me: if we have the defcustom for the "extra" does it make sense for it to include `'bibliography` as an option, so that if chosen, it's then automatically updated?
> 
> So then the defcustom would remove "extra" from the name?

That seems like the better options and `'bibliography`can be the default.

Anyhow I will be mostly absent this week. But if the design is settled and you (or someone esle) hasn't made the changes, I will make them over the weekend.

Reply:
> That seems like the better options and 'bibliography can be the default.

I think I would initially still keep it nil, so that users would need to explicitly opt in for the watching.

But I don't have a strong opinion.

> Anyhow I will be mostly absent this week. But if the design is settled and you (or someone esle) hasn't made the changes, I will make them over the weekend.

Sounds like a plan.

Reply:
BTW, I don't think this has any impact here, but FYI.

https://github.com/bdarcus/bibtex-actions/pull/222

Reply:
Also, I got some feedback that suggests we might want a defcustom for the notify function, with invalidating the cache being default?

Reply:
This now done along with independent local and global refreshes. Quite a lot of changes but in my testing everything is working fine. Though more testing is needed.

Reply:
> I think I would initially still keep it nil, so that users would need to explicitly opt in for the watching.

I have gone with the default being `'(bibliography)` because otherwise the user would be adding a line to their config which doesn't do anything unless they also set another variable. Which seems not ideal to me. 
Though I think `'(bibliography library notes)` is also a good default.

Reply:
> Also, I got some feedback that suggests we might want a defcustom for the notify function, with invalidating the cache being default?

What was the feedback? I have gone with it as the default but I personally prefer the `'refresh-cache` option. I so think providing both options is a good thing.

Reply:
That's fine. We can change it if people complain.

The feedback was a concern about Emacs blocking.

Reply:
> The feedback was a concern about Emacs blocking.

That seems a valid concern. I actually though about making refreshes async, but it seemed tricky without introducing races but doable. My own refreshes aren't frequent enough to make really want to do it.